<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Usage &mdash; CIAlign 1.1.0 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/sphinx_highlight.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="prev" title="Installation" href="installation.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html" class="icon icon-home"> CIAlign
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="introduction.html">CIAlign</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Usage</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#input-files">Input Files</a></li>
<li class="toctree-l2"><a class="reference internal" href="#quick-start">Quick Start</a></li>
<li class="toctree-l2"><a class="reference internal" href="#specifying-options">Specifying Options</a></li>
<li class="toctree-l2"><a class="reference internal" href="#basic-parameters">Basic Parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="#cleaning-functions">Cleaning Functions</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#remove-divergent">Remove Divergent</a></li>
<li class="toctree-l3"><a class="reference internal" href="#remove-insertions">Remove Insertions</a></li>
<li class="toctree-l3"><a class="reference internal" href="#crop-ends">Crop Ends</a></li>
<li class="toctree-l3"><a class="reference internal" href="#remove-short">Remove Short</a></li>
<li class="toctree-l3"><a class="reference internal" href="#remove-gap-only">Remove Gap Only</a></li>
<li class="toctree-l3"><a class="reference internal" href="#crop-divergent">Crop Divergent</a></li>
<li class="toctree-l3"><a class="reference internal" href="#retain">Retain</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#visualisation-functions">Visualisation functions</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#mini-alignments">Mini Alignments</a></li>
<li class="toctree-l3"><a class="reference internal" href="#sequence-logos">Sequence logos</a></li>
<li class="toctree-l3"><a class="reference internal" href="#statistics-plots">Statistics Plots</a></li>
<li class="toctree-l3"><a class="reference internal" href="#palettes">Palettes</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#interpretation-functions">Interpretation Functions</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#consensus-sequences">Consensus Sequences</a></li>
<li class="toctree-l3"><a class="reference internal" href="#position-frequency-probability-and-weight-matrices">Position Frequency, Probability and Weight Matrices</a></li>
<li class="toctree-l3"><a class="reference internal" href="#similarity-matrices">Similarity Matrices</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#additional-functions">Additional Functions</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#extracting-sections-from-an-alignment">Extracting sections from an alignment</a></li>
<li class="toctree-l3"><a class="reference internal" href="#replacing-u-or-t">Replacing U or T</a></li>
<li class="toctree-l3"><a class="reference internal" href="#unaligning-removing-gaps">Unaligning (removing gaps)</a></li>
</ul>
</li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">CIAlign</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a></li>
      <li class="breadcrumb-item active">Usage</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/pages/usage.md.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="usage">
<h1>Usage<a class="headerlink" href="#usage" title="Permalink to this heading">¶</a></h1>
<ol class="simple">
<li><p><a class="reference external" href="#input-files">Input Files</a></p></li>
<li><p><a class="reference external" href="#quick-start">Quick Start</a></p></li>
<li><p><a class="reference external" href="#specifying-options">Options</a></p></li>
<li><p><a class="reference external" href="#basic-parameters">Basic Parameters</a></p></li>
<li><p><a class="reference external" href="#cleaning-functions">Cleaning Functions</a></p></li>
<li><p><a class="reference external" href="#visualisation-functions">Visualisation Functions</a></p></li>
<li><p><a class="reference external" href="#interpretation-functions">Interpretation functions</a></p></li>
<li><p><a class="reference external" href="#editing-functions">Editing functions</a></p></li>
</ol>
<p>CIAlign is used to process multiple sequence alignments (MSAs) - sets of nucleotide or amino acid sequences which have already been aligned with an external tool.</p>
<section id="input-files">
<h2>Input Files<a class="headerlink" href="#input-files" title="Permalink to this heading">¶</a></h2>
<p>The input for CIAlign is an aligned MSA in FASTA format.</p>
</section>
<section id="quick-start">
<h2>Quick Start<a class="headerlink" href="#quick-start" title="Permalink to this heading">¶</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">CIAlign</span> <span class="pre">--infile</span> <span class="pre">INFILE</span> <span class="pre">--outfile_stem</span> <span class="pre">STEM</span> <span class="pre">OPTIONS</span></code></p>
<p>Where INFILE is the FASTA file you would like to process, STEM is a prefix for the output files and OPTIONS lists the functions you would like to run and the parameters you would like to use.</p>
<p>For example, to run the remove insertions function, with default settings, on the file <code class="docutils literal notranslate"><span class="pre">example1.fasta</span></code> and generate <code class="docutils literal notranslate"><span class="pre">ri_clean.fasta</span></code>.</p>
<p><code class="docutils literal notranslate"><span class="pre">CIAlign</span> <span class="pre">--infile</span> <span class="pre">example1.fasta</span> <span class="pre">--outfile_stem</span> <span class="pre">ri</span> <span class="pre">--remove_insertions</span></code></p>
<p>To run all functions with the default settings (please use this option cautiously):</p>
<p><code class="docutils literal notranslate"><span class="pre">CIAlign</span> <span class="pre">--infile</span> <span class="pre">example1.fasta</span> <span class="pre">--all</span></code></p>
</section>
<section id="specifying-options">
<h2>Specifying Options<a class="headerlink" href="#specifying-options" title="Permalink to this heading">¶</a></h2>
<p>Parameters can be specified in the command line OPTIONS or in a config file.</p>
<p>A template config file is provided in <code class="docutils literal notranslate"><span class="pre">CIAlign/templates/ini_template.ini</span></code> - edit this file and provide the path to the <code class="docutils literal notranslate"><span class="pre">--inifile</span></code> argument.</p>
<p><code class="docutils literal notranslate"><span class="pre">CIAlign</span> <span class="pre">--infile</span> <span class="pre">INFILE</span> <span class="pre">--outfile_stem</span> <span class="pre">STEM</span> <span class="pre">--inifile</span> <span class="pre">my_inifile.ini</span></code></p>
<p>If this argument is not provided command line arguments and defaults will be used.</p>
<p>Parameters passed in the command line will take precedence over config file parameters, which take precedence over defaults.</p>
<p>Command help can be accessed by typing <code class="docutils literal notranslate"><span class="pre">CIAlign</span> <span class="pre">--help</span></code></p>
</section>
<section id="basic-parameters">
<h2>Basic Parameters<a class="headerlink" href="#basic-parameters" title="Permalink to this heading">¶</a></h2>
<table border="1" class="docutils">
<thead>
<tr>
<th>Parameter</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--infile</code></td>
<td>Path to input alignment file in FASTA format</td>
<td>None</td>
</tr>
<tr>
<td><code>--inifile</code></td>
<td>Path to config file</td>
<td>None</td>
</tr>
<tr>
<td><code>--outfile_stem</code></td>
<td>Prefix for output files, including the path to the output directory</td>
<td>CIAlign</td>
</tr>
<tr>
<td><code>--all</code></td>
<td>Use all available cleaning, visualisation and interpretation functions with default parameters</td>
<td>False</td>
</tr>
<tr>
<td><code>--clean</code></td>
<td>Use all available cleaning functions with default parameters</td>
<td>False</td>
</tr>
<tr>
<td><code>--visualise</code></td>
<td>Use all available mini alignment visualisation functions with default parameters</td>
<td>False</td>
</tr>
<tr>
<td><code>--interpret</code></td>
<td>Use all available interpretation functions with default parameters</td>
<td>False</td>
</tr>
<tr>
<td><code>--silent</code></td>
<td>Do not print progress to the screen</td>
<td>False</td>
</tr>
<tr>
<td><code>--help</code></td>
<td>Show all available parameters with an explanation</td>
<td>None</td>
</tr>
<tr>
<td><code>--version</code></td>
<td>Show the version</td>
<td>None</td>
</tr>
</tbody>
</table><p>Beside these main parameters, the use of every function and corresponding thresholds can be specified by the user by adding parameters to the command line or by setting them in the configuration file. Available functions and their parameters are specified below.</p>
<p>CIAlign always produces a log file, specifying which functions have been run with witch parameters and what has been removed. It also outputs a machine parsable file that only specifies what has been removed with the original column positions and the sequence names.</p>
<p>Output files:</p>
<ul class="simple">
<li><p><strong><code class="docutils literal notranslate"><span class="pre">OUTFILE_STEM_log.txt</span></code></strong> - general log file</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">OUTFILE_STEM_removed.txt</span></code></strong> - removed columns positions and sequence names text file</p></li>
</ul>
</section>
<section id="cleaning-functions">
<h2>Cleaning Functions<a class="headerlink" href="#cleaning-functions" title="Permalink to this heading">¶</a></h2>
<p>The CIAlign cleaning functions are designed to address several common issues with multiple sequence alignments, affecting the speed, complexity and reliability of specific downstream analyses.</p>
<p>All of these functions remove columns or rows from the alignment to address sources of noise.</p>
<ul class="simple">
<li><p><a class="reference external" href="#remove-divergent">Remove divergent</a></p></li>
<li><p><a class="reference external" href="#remove-insertions">Remove insertions</a></p></li>
<li><p><a class="reference external" href="#crop-ends">Crop ends</a></p></li>
<li><p><a class="reference external" href="#remove-short">Remove short</a></p></li>
<li><p><a class="reference external" href="#remove-gap-only">Remove gap only</a></p></li>
<li><p><a class="reference external" href="#crop-divergent">Crop divergent</a></p></li>
</ul>
<p>Each of these steps (if specified) will be performed sequentially in the order specified in the table below.</p>
<p><code class="docutils literal notranslate"><span class="pre">remove_divergent</span></code>, <code class="docutils literal notranslate"><span class="pre">remove_insertions</span></code>, <code class="docutils literal notranslate"><span class="pre">crop_ends</span></code> and <code class="docutils literal notranslate"><span class="pre">crop</span> <span class="pre">divergent</span></code> require three or more sequences in the alignment, <code class="docutils literal notranslate"><span class="pre">remove_short</span></code> and <code class="docutils literal notranslate"><span class="pre">remove_gap_only</span></code> require two or more sequences.</p>
<p>Output files:</p>
<p>The “cleaned” alignment after all steps have been performed will be saved as <strong><code class="docutils literal notranslate"><span class="pre">OUTFILE_STEM_cleaned.fasta</span></code></strong></p>
<p>The <a class="reference external" href="#retain">retain</a> functions allow the user to specify sequences to keep regardless of the CIAlign results.</p>
<section id="remove-divergent">
<h3>Remove Divergent<a class="headerlink" href="#remove-divergent" title="Permalink to this heading">¶</a></h3>
<p>Removes divergent (outlier) sequences from the alignment. It is very common for an MSA to include one or a few outlier sequences which do not align well with the majority of the alignment. For some applications it is useful to remove these.</p>
<p>The remove divergent function specifically removes sequences with &lt;= <code class="docutils literal notranslate"><span class="pre">remove_divergent_minperc</span></code> positions at which the most common residue in the alignment is present.</p>
<table border="1" class="docutils">
<thead>
<tr>
<th>Parameter</th>
<th>Description</th>
<th>Default Value</th>
<th>Min</th>
<th>Max</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong><code>--remove_divergent</code></strong></td>
<td>Remove sequences with &lt;= <code>remove_divergent_minperc</code> positions at which the most common base / amino acid in the alignment is present</td>
<td>False</td>
<td>NA</td>
<td>NA</td>
</tr>
<tr>
<td><em><code>--remove_divergent_minperc</code></em></td>
<td>Minimum proportion of positions which should be identical to the most common base / amino acid in order to be preserved</td>
<td>remove_divergent_minperc_def</td>
<td>remove_divergent_minperc_min</td>
<td>remove_divergent_minperc_max</td>
</tr>
<tr>
<td><em><code>--remove_divergent_retain</code></em></td>
<td>Do not remove sequences with this name when running the remove divergent function</td>
<td>None</td>
<td>NA</td>
<td>NA</td>
</tr>
<tr>
<td><em><code>--remove_divergent_retain_str</code></em></td>
<td>Do not remove sequences with names containing this character string when running the remove divergent function</td>
<td>None</td>
<td>NA</td>
<td>NA</td>
</tr>
<tr>
<td><em><code>--remove_divergent_retain_list</code></em></td>
<td>Do not remove sequences with names listed in this file when running the remove divergent function</td>
<td>None</td>
<td>NA</td>
<td>NA</td>
</tr>
</tbody>
</table></section>
<section id="remove-insertions">
<h3>Remove Insertions<a class="headerlink" href="#remove-insertions" title="Permalink to this heading">¶</a></h3>
<p>Removes insertions which are not present in the majority of sequences (or regions which are deleted in the majority of sequences).  Insertions or other stretches of sequence which are only present in a minority of sequences can lead to large gaps, these are sometimes of interest but can also complicate downstream analysis.</p>
<p>The remove insertions function removes regions from the alignment which are found in &lt;= <code class="docutils literal notranslate"><span class="pre">insertion_min_perc</span></code> of the sequences but are surrounded by &gt;= <code class="docutils literal notranslate"><span class="pre">insertion_min_flank</span></code> columns of higher coverage.</p>
<table border="1" class="docutils">
<thead>
<tr>
<th>Parameter</th>
<th>Description</th>
<th>Default Value</th>
<th>Min</th>
<th>Max</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong><code>--remove_insertions</code></strong></td>
<td>Remove insertions found in &lt;= <code>insertion_min_perc</code> of sequences from the alignment</td>
<td>False</td>
<td>NA</td>
<td>NA</td>
</tr>
<tr>
<td><em><code>--insertion_min_size</code></em></td>
<td>Only remove insertions &gt;= this number of residues</td>
<td>insertion_min_size_def</td>
<td>insertion_min_size_min</td>
<td>insertion_min_size_max</td>
</tr>
<tr>
<td><em><code>--insertion_max_size</code></em></td>
<td>Only remove insertions &lt;= this number of residues</td>
<td>insertion_max_size_def</td>
<td>insertion_max_size_min</td>
<td>insertion_max_size_max</td>
</tr>
<tr>
<td><em><code>--insertion_min_flank</code></em></td>
<td>Minimum number of bases on either side of an insertion to classify it as an insertion</td>
<td>insertion_min_flank_def</td>
<td>insertion_min_flank_min</td>
<td>insertion_min_flank_max</td>
</tr>
<tr>
<td><em><code>--insertion_min_perc</code></em></td>
<td>Remove insertions which are present in less than this proportion of sequences</td>
<td>insertion_min_perc_def</td>
<td>insertion_min_perc_min</td>
<td>insertion_min_perc_max</td>
</tr>
</tbody>
</table></section>
<section id="crop-ends">
<h3>Crop Ends<a class="headerlink" href="#crop-ends" title="Permalink to this heading">¶</a></h3>
<p>It is common for an MSA to contain more gaps towards either end than in the body of the alignment, due to (for example) increased sequencing error towards the ends of reads, lower read coverage or assembly issues.</p>
<p>The crop ends function crops the ends of individual sequences if they contain a high proportion of gaps relative to the rest of the alignment. The number of gap positions separating every two consecutive non-gap positions at either end of the sequence is compared to a threshold (calculated from the total sequence length using <code class="docutils literal notranslate"><span class="pre">crop_ends_mingap_perc</span></code>) and if that difference is higher than the threshold, the start of the sequence will be reset to that position.</p>
<table border="1" class="docutils">
<thead>
<tr>
<th>Parameter</th>
<th>Description</th>
<th>Default Value</th>
<th>Min</th>
<th>Max</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong><code>--crop_ends</code></strong></td>
<td>Crop the ends of sequences if they are poorly aligned</td>
<td>False</td>
<td>NA</td>
<td>NA</td>
</tr>
<tr>
<td><em><code>--crop_ends_mingap_perc</code></em></td>
<td>Minimum proportion of the sequence length (excluding gaps) that is the threshold for change in gap numbers.</td>
<td>crop_ends_mingap_perc_def</td>
<td>crop_ends_mingap_perc_min</td>
<td>crop_ends_mingap_perc_max</td>
</tr>
<tr>
<td><em><code>--crop_ends_redefine_perc</code></em></td>
<td>Proportion of the sequence length (excluding gaps) that is being checked for change in gap numbers to redefine start/end.</td>
<td>crop_ends_redefine_perc_def</td>
<td>crop_ends_redefine_perc_min</td>
<td>crop_ends_redefine_perc_max</td>
</tr>
<tr>
<td><em><code>--crop_ends_retain</code></em></td>
<td>Do not crop sequences with this name when running the crop ends function</td>
<td>None</td>
<td>NA</td>
<td>NA</td>
</tr>
<tr>
<td><em><code>--crop_ends_retain_str</code></em></td>
<td>Do not crop sequences with names containing this character string when running the crop ends function</td>
<td>None</td>
<td>NA</td>
<td>NA</td>
</tr>
<tr>
<td><em><code>--crop_ends_retain_list</code></em></td>
<td>Do not crop sequences with names listed in this file when running the crop ends function</td>
<td>None</td>
<td>NA</td>
<td>NA</td>
</tr>
</tbody>
</table><p>Note: if the sequences are short (e.g. &lt; 100), a low crop_ends_mingap_perc (e.g. 0.01) will result in a change of gap numbers that is too low (e.g. 0). If this happens, the change in gap numbers will be set to 2 and a warning will be printed.</p>
</section>
<section id="remove-short">
<h3>Remove Short<a class="headerlink" href="#remove-short" title="Permalink to this heading">¶</a></h3>
<p>Removes short sequences below a threshold length.</p>
<table border="1" class="docutils">
<thead>
<tr>
<th>Parameter</th>
<th>Description</th>
<th>Default Value</th>
<th>Min</th>
<th>Max</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong><code>--remove_short</code></strong></td>
<td>Remove sequences &lt;= <code>remove_min_length</code>  amino acids from the alignment</td>
<td>False</td>
<td>NA</td>
<td>NA</td>
</tr>
<tr>
<td><em><code>--remove_min_length</code></em></td>
<td>Sequences are removed if they are shorter than this minimum length, excluding gaps.</td>
<td>remove_min_length_def</td>
<td>remove_min_length_min</td>
<td>remove_min_length_max</td>
</tr>
<tr>
<td><em><code>--remove_short_retain</code></em></td>
<td>Do not remove sequences with this name when running the remove short function</td>
<td>None</td>
<td>NA</td>
<td>NA</td>
</tr>
<tr>
<td><em><code>--remove_short_retain_str</code></em></td>
<td>Do not remove sequences with names containing this character string when running the remove short function</td>
<td>None</td>
<td>NA</td>
<td>NA</td>
</tr>
<tr>
<td><em><code>--remove_short_retain_list</code></em></td>
<td>Do not remove sequences with names listed in this file when running the remove short function</td>
<td>None</td>
<td>NA</td>
<td>NA</td>
</tr>
</tbody>
</table></section>
<section id="remove-gap-only">
<h3>Remove Gap Only<a class="headerlink" href="#remove-gap-only" title="Permalink to this heading">¶</a></h3>
<p>Removes columns containing only gaps. This function is run by default, to not run this function specify <code class="docutils literal notranslate"><span class="pre">--keep_gaponly</span></code>.</p>
<table border="1" class="docutils">
<thead>
<tr>
<th>Parameter</th>
<th>Description</th>
<th>Default Value</th>
<th>Min</th>
<th>Max</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong><code>--keep_gaponly</code></strong></td>
<td>Keep gap only columns in the alignment</td>
<td>False</td>
<td>NA</td>
<td>NA</td>
</tr>
</tbody>
</table></section>
<section id="crop-divergent">
<h3>Crop Divergent<a class="headerlink" href="#crop-divergent" title="Permalink to this heading">¶</a></h3>
<p>Some alignments have a region which is clearly of higher quality than the surrounding alignment, with less diversity and fewer gaps. This can be the case when regions have been extracted, e.g. from a full genome, but the start and end positions of the region of interest are not well defined·</p>
<p>The crop divergent function redefines the start and end positions of an alignment by looking for <code class="docutils literal notranslate"><span class="pre">crop_divergent_buffer_size</span></code> consecutive columns which have a minimum proportion of identical residues &gt;= <code class="docutils literal notranslate"><span class="pre">crop_divergent_min_prop_ident</span></code> and a minimum proportion of non-gap residues &gt;= <code class="docutils literal notranslate"><span class="pre">crop_divergent_min_prop_nongap</span></code>, then taking the first or last such column as the alignment start or end.</p>
<table border="1" class="docutils">
<thead>
<tr>
<th>Parameter</th>
<th>Description</th>
<th>Default Value</th>
<th>Min</th>
<th>Max</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong><code>--crop_divergent</code></strong></td>
<td>Crop either end of the alignment until &gt; <code>crop_divergent_min_prop_ident</code> residues in a column are identical and &gt; <code>crop_divergent_min_prop_nongap</code> residues are not gaps, over <code>buffer_size</code> consecutive columns</td>
<td>False</td>
<td>NA</td>
<td>NA</td>
</tr>
<tr>
<td><em><code>--crop_divergent_min_prop_ident</code></em></td>
<td>Minumum proportion of identical residues in a column to be retained by crop_divergent</td>
<td>divergent_min_prop_ident_def</td>
<td>divergent_min_prop_ident_min</td>
<td>divergent_min_prop_ident_max</td>
</tr>
<tr>
<td><em><code>--crop_divergent_min_prop_nongap</code></em></td>
<td>Minumum proportion of non gap residues in a column to be retained by crop_divergent</td>
<td>divergent_min_prop_nongap_def</td>
<td>divergent_min_prop_nongap_min</td>
<td>divergent_min_prop_nongap_max</td>
</tr>
<tr>
<td><em><code>--crop_divergent_buffer_size</code></em></td>
<td>Minumum number of consecutive columns which must meet the criteria for crop_divergent to be retained</td>
<td>divergent_buffer_size_def</td>
<td>divergent_buffer_size_min</td>
<td>divergent_buffer_size_max</td>
</tr>
</tbody>
</table></section>
<section id="retain">
<h3>Retain<a class="headerlink" href="#retain" title="Permalink to this heading">¶</a></h3>
<p>These parameters allow the user to specify sequences which should not be removed from the alignment.</p>
<p>The sequences can be specified by providing one or more sequence names (<code class="docutils literal notranslate"><span class="pre">--retain</span></code>), a character string to match in the names (<code class="docutils literal notranslate"><span class="pre">--retain_str</span></code>) or a file containing a list of names (<code class="docutils literal notranslate"><span class="pre">--retain_list</span></code>).</p>
<p>The crop ends, remove divergent and remove short functions also have the option to specify sequence names to ignore with those specific functions only.</p>
<table border="1" class="docutils">
<thead>
<tr>
<th>Parameter</th>
<th>Description</th>
<th>Default Value</th>
<th>Min</th>
<th>Max</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong><code>--retain</code></strong></td>
<td>Do not edit or remove sequences with this name when running any rowwise function (currently remove divergent, crop ends and remove short)</td>
<td>None</td>
<td>NA</td>
<td>NA</td>
</tr>
<tr>
<td><strong><code>--retain_str</code></strong></td>
<td>Do not edit or remove sequences with names containing this character string when running any rowwise function</td>
<td>None</td>
<td>NA</td>
<td>NA</td>
</tr>
<tr>
<td><strong><code>--retain_list</code></strong></td>
<td>Do not edit or remove sequences with names listed in this file when running any rowwise function</td>
<td>None</td>
<td>NA</td>
<td>NA</td>
</tr>
</tbody>
</table></section>
</section>
<section id="visualisation-functions">
<h2>Visualisation functions<a class="headerlink" href="#visualisation-functions" title="Permalink to this heading">¶</a></h2>
<p>Each of these functions produces some kind of visualisation of an MSA.</p>
<section id="mini-alignments">
<h3>Mini Alignments<a class="headerlink" href="#mini-alignments" title="Permalink to this heading">¶</a></h3>
<p>These functions produce “mini alignments” - images showing a small representation of your whole alignment, so that gaps and poorly aligned regions are clearly visible.
<img alt="Mini alignment example" src="../_images/mini_small.png" /></p>
<p>Output files:</p>
<ul class="simple">
<li><p><strong><code class="docutils literal notranslate"><span class="pre">OUTFILE_STEM_input.png</span> <span class="pre">(or</span> <span class="pre">svg,</span> <span class="pre">tiff,</span> <span class="pre">jpg)</span></code></strong> - visualisation of the input alignment</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">OUTFILE_STEM_output.png</span> <span class="pre">(or</span> <span class="pre">svg,</span> <span class="pre">tiff,</span> <span class="pre">jpg)</span></code></strong> -  visualisation of the cleaned output alignment</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">OUTFILE_STEM_markup.png</span> <span class="pre">(or</span> <span class="pre">svg,</span> <span class="pre">tiff,</span> <span class="pre">jpg)</span></code></strong> - visualisation of the input alignment with deleted rows and columns marked</p></li>
</ul>
<table border="1" class="docutils">
<thead>
<tr>
<th>Parameter</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong><code>--plot_input</code></strong></td>
<td>Plot a mini alignment - an image representing the input alignment</td>
<td>False</td>
</tr>
<tr>
<td><strong><code>--plot_output</code></strong></td>
<td>Plot a mini alignment - an image representing the output alignment</td>
<td>False</td>
</tr>
<tr>
<td><strong><code>--plot_markup</code></strong></td>
<td>Draws the input alignment but with the columns and rows which have been removed by each function marked up in corresponding colours</td>
<td>False</td>
</tr>
<tr>
<td><em><code>--plot_dpi</code></em></td>
<td>DPI for mini alignments</td>
<td>300</td>
</tr>
<tr>
<td><em><code>--plot_format</code></em></td>
<td>Image format for mini alignments - can be png, svg, tiff or jpg</td>
<td>png</td>
</tr>
<tr>
<td><em><code>--plot_width</code></em></td>
<td>Mini alignment width in inches</td>
<td>5</td>
</tr>
<tr>
<td><em><code>--plot_height</code></em></td>
<td>Mini alignment height in inches</td>
<td>3</td>
</tr>
<tr>
<td><em><code>--plot_keep_numbers</code></em></td>
<td>Label rows in mini alignments based on input alignment, rather than renumbering</td>
<td>False</td>
</tr>
<tr>
<td><em><code>--plot_force_numbers</code></em></td>
<td>Force all rows in mini alignments to be numbered rather than labelling e.g. every 10th row for larger plots Will cause labels to overlap in larger plots</td>
<td>False</td>
</tr>
</tbody>
</table></section>
<section id="sequence-logos">
<h3>Sequence logos<a class="headerlink" href="#sequence-logos" title="Permalink to this heading">¶</a></h3>
<p>These functions draw sequence logos representing output (cleaned) alignment using the algorithm specified by [Schneider (1990)(]https://www.ncbi.nlm.nih.gov/pmc/articles/PMC332411/).</p>
<p>Traditional “text” sequence logos can be produced as well as bar charts summarising the same information.</p>
<p><strong>Text</strong></p>
<p><img alt="Text logo example" src="../_images/sequence_logo_small.png" /></p>
<p><strong>Bar</strong></p>
<p><img alt="Bar logo example" src="../_images/sequence_logo_bar_small.png" /></p>
<p>You can also specify a subsection of the alignment using the <code class="docutils literal notranslate"><span class="pre">logo_start</span></code> and <code class="docutils literal notranslate"><span class="pre">logo_end</span></code> arguments, positions should be relative to the input alignment. If no cleaning functions are specified, the logo will be based on your input alignment.</p>
<p>Output_files:</p>
<ul class="simple">
<li><p><strong><code class="docutils literal notranslate"><span class="pre">OUTFILE_STEM_logo_bar.png</span> <span class="pre">(or</span> <span class="pre">svg,</span> <span class="pre">tiff,</span> <span class="pre">jpg)</span></code></strong> - the alignment represented as a bar chart</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">OUTFILE_STEM_logo_text.png</span> <span class="pre">(or</span> <span class="pre">svg,</span> <span class="pre">tiff,</span> <span class="pre">jpg)</span></code></strong> - the alignment represented as a standard sequence logo using text</p></li>
</ul>
<table border="1" class="docutils">
<thead>
<tr>
<th>Parameter</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong><code>--make_sequence_logo</code></strong></td>
<td>Draw a sequence logo</td>
<td>False</td>
</tr>
<tr>
<td><em><code>--sequence_logo_type</code></em></td>
<td>Type of sequence logo - bar/text/both</td>
<td>bar</td>
</tr>
<tr>
<td><em><code>--sequence_logo_dpi</code></em></td>
<td>DPI for sequence logo</td>
<td>300</td>
</tr>
<tr>
<td><em><code>--sequence_logo_font</code></em></td>
<td>Font (see NB below) for bases / amino acids in a text based sequence logo</td>
<td>monospace</td>
</tr>
<tr>
<td><em><code>--sequence_logo_nt_per_row</code></em></td>
<td>Number of bases / amino acids to show per row in the sequence logo, where the logo is too large to show on a single line</td>
<td>50</td>
</tr>
<tr>
<td><em><code>--sequence_logo_filetype</code></em></td>
<td>Image file type to use for the sequence logo - can be png, svg, tiff or jpg</td>
<td>png</td>
</tr>
<tr>
<td><em><code>--logo_start</code></em></td>
<td>Start sequence logo</td>
<td>0</td>
</tr>
<tr>
<td><em><code>--logo_end</code></em></td>
<td>End of sequence logo</td>
<td>MSA length</td>
</tr>
</tbody>
</table><p>NB: to see available fonts on your system, run CIAlign –list_fonts_only and view CIAlign_fonts.png</p>
</section>
<section id="statistics-plots">
<h3>Statistics Plots<a class="headerlink" href="#statistics-plots" title="Permalink to this heading">¶</a></h3>
<p>For each position in the alignment, these functions plot:</p>
<ul class="simple">
<li><p>Coverage (the number of non-gap residues)</p></li>
<li><p>Information content</p></li>
<li><p>Shannon entropy</p></li>
</ul>
<p><strong>Coverage</strong></p>
<p><img alt="../_images/cov_small.png" src="../_images/cov_small.png" /></p>
<p><strong>Information Content</strong></p>
<p><img alt="../_images/ic_small.png" src="../_images/ic_small.png" /></p>
<p><strong>Shannon Entropy</strong></p>
<p><img alt="../_images/se_small.png" src="../_images/se_small.png" /></p>
<p>Output files:</p>
<ul class="simple">
<li><p><strong><code class="docutils literal notranslate"><span class="pre">OUTFILE_STEM_input_coverage.png</span> <span class="pre">(or</span> <span class="pre">svg,</span> <span class="pre">tiff,</span> <span class="pre">jpg)</span> </code></strong> - image showing the input alignment coverage</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">OUTFILE_STEM_output_coverage.png</span> <span class="pre">(or</span> <span class="pre">svg,</span> <span class="pre">tiff,</span> <span class="pre">jpg)</span> </code></strong> - image showing the output alignment coverage</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">OUTFILE_STEM_input_information_content.png</span> <span class="pre">(or</span> <span class="pre">svg,</span> <span class="pre">tiff,</span> <span class="pre">jpg)</span> </code></strong> - image showing the input alignment information content</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">OUTFILE_STEM_output_information_content.png</span> <span class="pre">(or</span> <span class="pre">svg,</span> <span class="pre">tiff,</span> <span class="pre">jpg)</span> </code></strong> - image showing the output alignment information content</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">OUTFILE_STEM_input_shannon_entropy.png</span> <span class="pre">(or</span> <span class="pre">svg,</span> <span class="pre">tiff,</span> <span class="pre">jpg)</span> </code></strong> - image showing the input alignment Shannon entropy</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">OUTFILE_STEM_output_shannon_entropy.png</span> <span class="pre">(or</span> <span class="pre">svg,</span> <span class="pre">tiff,</span> <span class="pre">jpg)</span> </code></strong> - image showing the output alignment Shannon entropy</p></li>
</ul>
<table border="1" class="docutils">
<thead>
<tr>
<th>Parameter</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong><code>--plot_stats_input</code></strong></td>
<td>Plot the statistics for the input MSA</td>
<td>False</td>
</tr>
<tr>
<td><strong><code>--plot_stats_output</code></strong></td>
<td>Plot the statistics for the output MSA</td>
<td>False</td>
</tr>
<tr>
<td><em><code>--plot_stats_dpi</code></em></td>
<td>DPI for coverage plot</td>
<td>300</td>
</tr>
<tr>
<td><em><code>--plot_stats_height</code></em></td>
<td>Height for coverage plot (inches)</td>
<td>3</td>
</tr>
<tr>
<td><em><code>--plot_stats_width</code></em></td>
<td>Width for coverage plot (inches)</td>
<td>5</td>
</tr>
<tr>
<td><em><code>--plot_stats_colour</code></em></td>
<td>Colour for coverage plot (hex code or name)</td>
<td>#007bf5</td>
</tr>
<tr>
<td><em><code>--plot_stats_filetype</code></em></td>
<td>File type for coverage plot (png, svg, tiff, jpg)</td>
<td>png</td>
</tr>
</tbody>
</table></section>
<section id="palettes">
<h3>Palettes<a class="headerlink" href="#palettes" title="Permalink to this heading">¶</a></h3>
<p>This function sets the colour palette for the mini alignments. Currently available palettes are colour blind safe (CBS) and bright.</p>
<p><strong>CBS</strong></p>
<p><img alt="CBS palette" src="../_images/CBS_small.png" /></p>
<p><strong>Bright</strong>
<img alt="Bright palette" src="../_images/bright_small.png" /></p>
<table border="1" class="docutils">
<thead>
<tr>
<th>Parameter</th>
<th>Description</th>
<th>Default Value</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong><code>--palette</code></strong></td>
<td>Colour palette. Currently implemented CBS (colourblind safe) and bright</td>
<td>CBS</td>
</tr>
</tbody>
</table></section>
</section>
<section id="interpretation-functions">
<h2>Interpretation Functions<a class="headerlink" href="#interpretation-functions" title="Permalink to this heading">¶</a></h2>
<p>These functions provide additional analyses you may wish to perform on your alignment.</p>
<section id="consensus-sequences">
<h3>Consensus Sequences<a class="headerlink" href="#consensus-sequences" title="Permalink to this heading">¶</a></h3>
<p>This step generates a consensus sequence based on the cleaned alignment.  If no cleaning functions are performed, the consensus will be based on the input alignment.
For the “majority” based consensus sequences, where the two most frequent characters are equally common a random character is selected.</p>
<p>Output files:</p>
<ul class="simple">
<li><p><strong><code class="docutils literal notranslate"><span class="pre">OUTFILE_STEM_consensus.fasta</span></code></strong> - the consensus sequence only</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">OUTFILE_STEM_with_consensus.fasta</span></code></strong> - the cleaned alignment plus the consensus</p></li>
</ul>
<table border="1" class="docutils">
<thead>
<tr>
<th>Parameter</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong><code>--make_consensus</code></strong></td>
<td>Make a consensus sequence based on the cleaned alignment</td>
<td>False</td>
</tr>
<tr>
<td><em><code>--consensus_type</code></em></td>
<td>Type of consensus sequence to make - can be majority, to use the most common character at each position in the consensus, even if this is a gap, or majority_nongap, to use the most common non-gap character at each position</td>
<td>majority</td>
</tr>
<tr>
<td><em><code>--consensus_keep_gaps</code></em></td>
<td>If there are gaps in the consensus (if majority_nongap is used as consensus_type), should these be included in the consensus (True) or should this position in the consensus be deleted (False)</td>
<td>False</td>
</tr>
<tr>
<td><em><code>--consensus_name</code></em></td>
<td>Name to use for the consensus sequence in the output fasta file</td>
<td>consensus</td>
</tr>
</tbody>
</table></section>
<section id="position-frequency-probability-and-weight-matrices">
<h3>Position Frequency, Probability and Weight Matrices<a class="headerlink" href="#position-frequency-probability-and-weight-matrices" title="Permalink to this heading">¶</a></h3>
<p>These functions are used to create a position weight matrix, position frequency matrix or position probability matrix for your input or output (cleaned) alignment. These are numerical representations of the alignment which can be used as input for various other software, for example to find regions of another sequence resembling part of your alignment. PFMs, PPMs and PWMs are described well in the Wikipedia article <a class="reference external" href="https://en.wikipedia.org/wiki/Position_weight_matrix">here</a>.</p>
<p>You can also specify a subsection of the alignment using the <code class="docutils literal notranslate"><span class="pre">pwm_start</span></code> and <code class="docutils literal notranslate"><span class="pre">pwm_end</span></code> arguments, positions should be relative to the input alignment.</p>
<p>Output_files:</p>
<ul class="simple">
<li><p><strong><code class="docutils literal notranslate"><span class="pre">OUTFILE_STEM_pwm_(input/output).txt</span></code></strong> - position weight matrix representing the alignment (or part of the alignment)</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">OUTFILE_STEM_ppm_(input/output).txt</span></code></strong> - position probability matrix representing the alignment (or part of the alignment)</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">OUTFILE_STEM_pfm_(input/output).txt</span></code></strong> - position frequency matrix representing the alignment (or part of the alignment)</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">OUTFILE_STEM_ppm_meme_(input/output).txt</span></code></strong> - position probability matrix representing the alignment (or part of the alignment) in the format used by the MEME software suite.</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">OUTFILE_STEM_blamm_(input/output).png</span></code></strong> - position probability matrix representing the alignment (or part of the alignment) in the format used by the BLAMM software tool.</p></li>
</ul>
<table border="1" class="docutils">
<thead>
<tr>
<th>Parameter</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong><code>--pwm_input</code></strong></td>
<td>Generate a position frequency matrix, position probability matrix and position weight matrix based on the input alignment</td>
<td>False</td>
</tr>
<tr>
<td><strong><code>--pwm_output</code></strong></td>
<td>Generate a position frequency matrix, position probability matrix and position weight matrix based on the cleaned output alignment</td>
<td>False</td>
</tr>
<tr>
<td><em><code>--pwm_start</code></em></td>
<td>Start the PWM and other matrices from this column of the input alignment</td>
<td>None</td>
</tr>
<tr>
<td><em><code>--pwm_end</code></em></td>
<td>Start the PWM and other matrices from this column of the input alignment</td>
<td>None</td>
</tr>
<tr>
<td><em><code>--pwm_freqtype</code></em></td>
<td>Type of background frequency matrix to use when generating the PWM. Should be 'equal', 'calc', 'calc2' or 'user'. 'equal', assume all residues are equally common, 'calc', frequency is calculated using the PFM,  'calc2', frequency is calculated using the full alignment (same as calc if pwm_start and pwm_end are not specified).</td>
<td>equal</td>
</tr>
<tr>
<td><em><code>--pwm_alphatype</code></em></td>
<td>Alpha value to use as a pseudocount to avoid zero values in the PPM. Should be 'calc' or 'user'. If alphatype is 'calc', alpha is calculated as frequency(base) * (square root(n rows in alignment)), as described in Dave Tang's blog <a href="https://davetang.org/muse/2013/10/01/position-weight-matrix/">here</a>, which recreates the method used in <a href="doi.org/10.1038/nrg1315">Wasserman &amp; Sandelin 2004</a>. If alpha type is 'user' the user provides the value of alpha as <code>pwm_alphatype</code>. To run without pseudocounts set <code>pwm_alphatype</code> as user and <code>pwm_alphaval</code> as 0</td>
<td>calc</td>
</tr>
<tr>
<td><em><code>--pwm_alphaval</code></em></td>
<td>User defined value of the alpha parameter to use as a pseudocount in the PPM.</td>
<td>1</td>
</tr>
<tr>
<td><em><code>--pwm_output_blamm</code></em></td>
<td>Output PPM formatted for <a href="https://github.com/biointec/blamm">BLAMM</a> software</td>
<td>False</td>
</tr>
<tr>
<td><em><code>--pwm_output_meme</code></em></td>
<td>Output PPM formatted for <a href="https://meme-suite.org/meme">MEME</a> software</td>
<td>False</td>
</tr>
</tbody>
</table></section>
<section id="similarity-matrices">
<h3>Similarity Matrices<a class="headerlink" href="#similarity-matrices" title="Permalink to this heading">¶</a></h3>
<p>Generates a matrix showing the proportion of identical bases / amino acids between each pair of sequences in the MSA.</p>
<p>Output file:</p>
<ul class="simple">
<li><p><strong><code class="docutils literal notranslate"><span class="pre">OUTFILE_STEM_input_similarity.tsv</span></code></strong> - similarity matrix for the input file</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">OUTFILE_STEM_output_similarity.tsv</span></code></strong> - similarity matrix for the output file</p></li>
</ul>
<table border="1" class="docutils">
<thead>
<tr>
<th>Parameter</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong><code>--make_similarity_matrix_input</code></strong></td>
<td>Make a similarity matrix for the input alignment</td>
<td>False</td>
</tr>
<tr>
<td><strong><code>--make_similarity_matrix_output</code></strong></td>
<td>Make a similarity matrix for the output alignment</td>
<td>False</td>
</tr>
<tr>
<td><em><code>--make_simmatrix_keepgaps</code></em></td>
<td>0 - exclude positions which are gaps in either or both sequences from similarity calculations, 1 - exclude positions which are gaps in both sequences, 2 - include all positions</td>
<td>0</td>
</tr>
<tr>
<td><em><code>--make_simmatrix_dp</code></em></td>
<td>Number of decimal places to display in the similarity matrix output file</td>
<td>4</td>
</tr>
<tr>
<td><em><code>--make_simmatrix_minoverlap</code></em></td>
<td>Minimum overlap between two sequences to have non-zero similarity in the similarity matrix</td>
<td>1</td>
</tr>
</tbody>
</table></section>
</section>
<section id="additional-functions">
<h2>Additional Functions<a class="headerlink" href="#additional-functions" title="Permalink to this heading">¶</a></h2>
<section id="extracting-sections-from-an-alignment">
<h3>Extracting sections from an alignment<a class="headerlink" href="#extracting-sections-from-an-alignment" title="Permalink to this heading">¶</a></h3>
</section>
<section id="replacing-u-or-t">
<h3>Replacing U or T<a class="headerlink" href="#replacing-u-or-t" title="Permalink to this heading">¶</a></h3>
<p>This function replaces the U nucleotides with T nucleotides or vice versa without otherwise changing the alignment.</p>
<p>Output files:</p>
<ul class="simple">
<li><p><strong><code class="docutils literal notranslate"><span class="pre">OUTFILE_STEM_T_input.fasta</span></code></strong> - input alignment with T’s instead of U’s</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">OUTFILE_STEM_T_output.fasta</span></code></strong> - output alignment with T’s instead of U’s</p></li>
</ul>
<p>or</p>
<ul class="simple">
<li><p><strong><code class="docutils literal notranslate"><span class="pre">OUTFILE_STEM_U_input.fasta</span></code></strong> - input alignment with U’s instead of T’s</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">OUTFILE_STEM_U_output.fasta</span></code></strong> - output alignment with U’s instead of T’s</p></li>
</ul>
<table border="1" class="docutils">
<thead>
<tr>
<th>Parameter</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--replace_input_tu</code></td>
<td>Generates a copy of the input alignment with T's instead of U's</td>
<td>False</td>
</tr>
<tr>
<td><code>--replace_output_tu</code></td>
<td>Generates a copy of the output alignment with T's instead of U's</td>
<td>False</td>
</tr>
<tr>
<td><code>--replace_input_ut</code></td>
<td>Generates a copy of the input alignment with U's instead of T's</td>
<td>False</td>
</tr>
<tr>
<td><code>--replace_output_ut</code></td>
<td>Generates a copy of the output alignment with U's instead of T's</td>
<td>False</td>
</tr>
</tbody>
</table></section>
<section id="unaligning-removing-gaps">
<h3>Unaligning (removing gaps)<a class="headerlink" href="#unaligning-removing-gaps" title="Permalink to this heading">¶</a></h3>
<p>This function simply removes the gaps from the input or output alignment and creates and unaligned file of the sequences.</p>
<p>Output files:</p>
<ul class="simple">
<li><p><strong><code class="docutils literal notranslate"><span class="pre">OUTFILE_STEM_unaligned_input.fasta</span></code></strong> - unaligned sequences of input alignment</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">OUTFILE_STEM_unaligned_output.fasta</span></code></strong> - unaligned sequences of output alignment</p></li>
</ul>
<table border="1" class="docutils">
<thead>
<tr>
<th>Parameter</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--unalign_input</code></td>
<td>Generates a copy of the input alignment with no gaps</td>
<td>False</td>
</tr>
<tr>
<td><code>--unalign_output</code></td>
<td>Generates a copy of the output alignment with no gaps</td>
<td>False</td>
</tr>
</tbody>
</table></section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="installation.html" class="btn btn-neutral float-left" title="Installation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Charlotte Tumescheit, Andrew Firth, Katy Brown.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>