### Full output tables for benchmarking analysis with simulated data and for using CIAlign with biological data. Column headings are described at the top of each file. Reference numbers within files correspond to numbering in the CIAlign manuscript.

### Supplementary Table 1
Table showing the results of 100 simulations using each of EvolvAGene [1] and INDELible [2] and cleaning with CIAlign with relaxed, moderate and stringent settings. Column headings are described at the top of the file.

### Supplementary Table 2
Table showing the comparison of consensus sequences generated from alignments of  the results of 100 simulations each of good quality, medium quality and poor quality Oxford Nanopore reads using BadRead [3] with the original input sequence. Column headings are described at the top of the file.

### Supplementary Table 3
Table showing the results of cleaning Pfam seed and full alignments [4] using CIAlign. Column headings are described at the top of the file.

### Supplementary Table 4
Table showing the results of cleaning FASTA files containing indels from the 1000 genomes project [5] using CIAlign. Column headings are described at the top of the file.

## References
1. Hall BG. Simulating DNA coding sequence evolution with EvolveAGene 3. Mol Biol Evol. 2008;25:688–95.
2. Fletcher W, Yang Z. INDELible: a flexible simulator of biological sequence evolution. Mol Biol Evol. 2009;26:1879–88.
3. Wick RR. Badread: simulation of error-prone long reads. J Open Source Softw. 2019;4:1316.
4. Finn RD, Bateman A, Clements J, Coggill P, Eberhardt RY, Eddy SR, et al. Pfam: the protein families database. Nucleic Acids Res. 2014;42 Database issue:D222–30.
5. Auton A, Abecasis GR, Altshuler DM, Durbin RM, Abecasis GR, Bentley DR, et al. A global reference for human genetic variation. Nature. 2015;526:68–74.