### Full output tables for benchmarking analysis with simulated data and for using CIAlign with biological data. Column headings are described at the top of each file.

### Online Table 1
Table showing the results of 100 simulations using each of EvolvAGene [1] and INDELible [2] and with the BAliBase reference alignments [3], cleaned with CIAlign with relaxed, moderate and stringent settings.

### Online Table 2
Table showing the comparison of consensus sequences generated from alignments of  the results of 100 simulations each of good quality, medium quality and poor quality Oxford Nanopore reads using BadRead [4] with the original input sequence.

### Online Table 3
Table showing the results of running CIAlign on the HomFam reference alignments [5].

### Online Table 4
Table showing the results of comparing the secondary protein structure for reference sequences provided with the QuanTest2 [6] tool and secondary structure for these reference sequences once incorporated into a test alignment, before and after running CIAlign.

### Online Table 5
Table showing the results of 100 simulations using each of EvolvAGene [1] and INDELible [2] and with the BAliBase reference alignments [3], cleaned with CIAlign with relaxed, moderate and stringent settings, using a large selection of different alignment tools.

### Online Table 6
Table showing the results of cleaning the simulated alignments using CIAlign with relaxed, moderate and stringent settings plus the results of using other tools GBlocks [7], TrimAl[8] and ZORRO [9].

### Online Table 7
Table showing the results of cleaning Pfam seed and full alignments [10] using CIAlign.

### Online Table 8
Table showing the results of cleaning FASTA files containing indels from the 1000 genomes project [11] using CIAlign.

## References
1. Hall BG. Simulating DNA coding sequence evolution with EvolveAGene 3. Mol Biol Evol. 2008;25:688–95.
2. Fletcher W, Yang Z. INDELible: a flexible simulator of biological sequence evolution. Mol Biol Evol. 2009;26:1879–88.
3. Bahr A, Thompson JD, Thierry J-C, Poch O. BAliBASE (Benchmark Alignment dataBASE): enhancements for repeats, transmembrane sequences and circular permutations. Nucleic Acids Res. 2001;29:323–6.
4. Wick RR. Badread: simulation of error-prone long reads. J Open Source Softw. 2019;4:1316.
5. Sievers F, Dineen D, Wilm A, Higgins DG. Making automated multiple alignments of very large numbers of protein sequences. Bioinformatics. 2013;29:989–95.
6. Sievers F, Higgins DG. QuanTest2: benchmarking multiple sequence alignments using secondary structure prediction. Bioinformatics. 2020;36:90–5.
7. Talavera G, Castresana J. Improvement of phylogenies after removing divergent and ambiguously aligned blocks from protein sequence alignments. Syst Biol. 2007;56:564–77.
8. Capella-Gutiérrez S, Silla-Martínez JM, Gabaldón T. trimAl: a tool for automated alignment trimming in large-scale phylogenetic analyses. Bioinformatics. 2009;25:1972–3.
9. Wu M, Chatterji S, Eisen JA. Accounting For Alignment Uncertainty in Phylogenomics. PLOS ONE. 2012;7:e30288.
10. Finn RD, Bateman A, Clements J, Coggill P, Eberhardt RY, Eddy SR, et al. Pfam: the protein families database. Nucleic Acids Res. 2014;42 Database issue:D222–30.
11. Auton A, Abecasis GR, Altshuler DM, Durbin RM, Abecasis GR, Bentley DR, et al. A global reference for human genetic variation. Nature. 2015;526:68–74.